<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="VarSim">
    <meta name="author" content="John C. Mu">

    <title>VarSim</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .container {
        margin-right: auto;
        margin-left: auto;
        max-width: 760px;
        }
    </style>

    <!-- Google Tag Manager -->
    <noscript>
        <iframe src="//www.googletagmanager.com/ns.html?id=GTM-WGQXX7"
                height="0" width="0" style="display:none;visibility:hidden"></iframe>
    </noscript>
    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'//www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','GTM-WGQXX7');
    </script>
    <!-- End Google Tag Manager -->

    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-58409498-2', 'auto');
  ga('require', 'linkid', 'linkid.js');
  ga('send', 'pageview');
  
  
  
  /**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}


function _gaLt(event){
    var el = event.srcElement || event.target;

    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }

    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );

            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }

    }
}

/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});


    </script>

</head>

<body>


<div class="container" style="padding-bottom: 400px;">
    <div class="page-header">
        <h1>VarSim</h1>

        <h2>A high-fidelity simulation validation framework for high-throughput genome sequencing with cancer
            applications</h2>
    </div>

    <div class="jumbotron">
        <p>
            <a href="webapp/alignment_compare.html"
               onclick="trackOutboundLink('webapp/alignment_compare.html'); return false;">Alignment Compare</a> webapp
        </p>

        <p>
            <a href="webapp/variant_compare.html"
               onclick="trackOutboundLink('webapp/variant_compare.html'); return false;">Variant Compare</a> webapp
        </p>

        <p>
            <i>For more information contact us at <a href="mailto:bina.rd@roche.com">bina.rd@roche.com</a></i>
        </p>

    </div>

    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="font-family:monospace;">
        <div class="panel-body">
            <i>If you use VarSim in your work, please cite the following:</i><br>
            John C. Mu, Marghoob Mohiyuddin, Jian Li, Narges Bani Asadi, Mark B. Gerstein, Alexej Abyzov, Wing H. Wong,
            and Hugo Y.K. Lam<br>
            <b>VarSim: A high-fidelity simulation and validation framework for high-throughput genome sequencing with
                cancer applications</b><br>
            Bioinformatics first published online December 17, 2014 <a
                href="http://dx.doi.org/10.1093/bioinformatics/btu828"
                onclick="trackOutboundLink('http://dx.doi.org/10.1093/bioinformatics/btu828'); return false;">doi:10.1093/bioinformatics/btu828</a>
        </div>
    </div>
    <h2>Data Access</h2>

    <p>
        Pre-generated reads and variants are avaliable on Amazon S3. It is in a requester-pays bucket
        so the requester must pay for the data transmission costs outside of S3. Currently the cost Amazon
        lists is 0.09 $/GB. You will also need an AWS account to access the data.
    </p>

    <div class="well well-sm" style="font-family:monospace;">
        Bucket URI: s3://varsim
    </div>
    <h3>Instructions on how to access the data</h3>
    <ol>
        <li>Create an account at aws.amazon.com</li>
        <li>Download an AWS S3 client, this example will use <code>http://s3tools.org/s3cmd</code></li>
        <li>It must be the beta version of <code>s3cmd</code> (1.53+), if you use <code>pip</code>, use the command
            <code>pip -vvv install s3cmd --pre</code></li>
        <li>Configure your s3cmd with <code>s3cmd configure</code>, get authentication details from <code>https://console.aws.amazon.com/iam/home?#security_credential</code>
            then "access keys"
        </li>
        <li>List files with the command <code>s3cmd ls --add-header=x-amz-request-payer:requester s3://varsim/*</code>
        </li>
        <li>Download files with the command <code>s3cmd get --add-header=x-amz-request-payer:requester s3://varsim/[file
            that you want]</code></li>
    </ol>
    <p>
        There is a <code>readme.txt</code> in the root directory with details of the data.
    </p>

    <h2>Download VarSim</h2>

    <p>Latest pre-built version and source code: <a
            href="https://github.com/bioinform/varsim/releases/"
            onclick="trackOutboundLink('https://github.com/bioinform/varsim/releases/'); return false;">https://github.com/bioinform/varsim/releases/</a>
    </p>

    <h2>Docker</h2>
    <p>Available at <a href="https://hub.docker.com/r/carlocreator/varsim">carlocreator/varsim</a>.Alternatively, try out <code>docker/test.sh</code> in VarSim repo.</p>

    <h2>VarSim on PrecisionFDA</h2>

    <p>VarSim is also available as an app on the <a href="https://precision.fda.gov/" onclick="trackOutboundLink('https://precision.fda.gov/'); return false;">PrecisionFDA</a> portal. If you have an account, please check out this <a href="https://precision.fda.gov/notes/15-varsim-app" onclick="trackOutboundLink('https://precision.fda.gov/notes/15-varsim-app'); return false;">note</a></p>


    <h2>System Requirements</h2>

    <p>
        For variant simulation and read generation/simulation:
    <ul>
        <li>32GB free RAM</li>
        <li>Enough free disk space to store twice the number of reads generated</li>
        <li>ART or dwgsim installed</li>
        <li>Currently only supports human genome simulation</li>
    </ul>
    </p>
    <p>
        For alignment and variant calling validation:
    <ul>
        <li>8GB free RAM</li>
    </ul>
    </p>

    <h2>Building VarSim</h2>

    <p>To build, execute <code>./build.sh</code></p>

    <h2>Running VarSim</h2>

    <p>Type <code>./varsim.py -h</code> for help.</p>

    <p><code>varsim_somatic.py</code> is a helper script for the somatic workflow.
        This is in its early stages. Please contact us if you have trouble with it</p>

    <h2>Quick Start Guide for Germline Simulation</h2>


    <p>This quick start guide will provide steps for generating a random genome with pre-specified and random variants.
        Then generate reads from this genome with ART. Finally, results of analysis on the output of secondary analysis
        is plotted.
    </p>
    <p><mark>
		20190410 update: some of the documentation might be obsolete, please run <code>./build.sh</code>
	    inside varsim folder to build varsim executables, and third-party software (e.g. ART, SAMtools), then
	    go to <code>tests/quickstart_test</code> and run <code>./quickstart.sh</code>
	    </mark>
	</p>

    <p>
        <b>Step 1:</b> Create a directory <code>varsim_run</code>
    </p>

    Download varsim to <code>varsim_run</code>

<pre>
wget https://github.com/bioinform/varsim/releases/download/v0.7.1/varsim-0.7.1.tar.gz
tar xfz varsim-0.7.1.tar.gz
</pre>

    <p>
        <b>Step 2:</b> Download the following files to <code>varsim_run</code>:
    </p>

    <ol>
        <li>Reference genome in FASTA format <a href='ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz'>[hs37d5.fa.gz]</a></li>
        <li>Concatented insert sequences as <a href='http://web.stanford.edu/group/wonglab/varsim/insert_seq.txt'>[insert_seq.txt]</a>
        </li>
        <li>DGV annotations <a
                href='http://web.stanford.edu/group/wonglab/varsim/GRCh37_hg19_supportingvariants_2013-07-23.txt'>[GRCh37_hg19_supportingvariants_2013-07-23.txt]</a>
        </li>
        <li>dbSNP 141 file <a href='ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b147_GRCh37p13/VCF/All_20160601.vcf.gz'>[All.vcf.gz]</a></li>
        <li>(optional) Extra variants to include in simulation variants.vcf</li>
    </ol>

    <p>
        by running the following commands
    </p>

<pre>
wget http://goo.gl/lgT18V
wget http://web.stanford.edu/group/wonglab/varsim/insert_seq.txt
wget http://web.stanford.edu/group/wonglab/varsim/GRCh37_hg19_supportingvariants_2013-07-23.txt
wget http://goo.gl/sq4mOE
gunzip hs37d5.fa.gz
</pre>

    <p>
        <b>Step 3:</b> Download Samtools to index the reference genome
    </p>

<pre>
mkdir samtools
cd samtools
wget http://sourceforge.net/projects/samtools/files/samtools/1.0/samtools-bcftools-htslib-1.0_x64-linux.tar.bz2
tar xfj samtools-bcftools-htslib-1.0_x64-linux.tar.bz2
cd ..
samtools/samtools-bcftools-htslib-1.0_x64-linux/bin/samtools faidx hs37d5.fa
</pre>

    <p>
        <b>Step 4:</b> Dowload ART to <code>varsim_run</code> under an the directory ART by running

    <p>

<pre>
mkdir ART
cd ART
wget http://www.niehs.nih.gov/research/resources/assets/docs/artbinvanillaicecream031114linux64tgz.tgz
tar xfz artbinvanillaicecream031114linux64tgz.tgz
cd ..
</pre>

    <p>
        The structure of <code>varsim_run</code> should look like
    </p>

<pre>
\varsim_run
- All.vcf.gz  
- \ART  
- - \art_bin_VanillaIceCream  
- - - ART files ...
- - artbinvanillaicecream031114linux64tgz.tgz
- GRCh37_hg19_supportingvariants_2013-07-23.txt	
- hs37d5.fa  
- hs37d5.fa.fai  
- insert_seq.txt  
- \samtools  
- - samtools files...
- VarSim.jar
- varsim.py
- varsim_somatic.py
- varsim-0.7.1.tar.gz
</pre>

    <p>
        <b>Step 5:</b> Run the following command to generate the simulated genome and reads to 30x depth.
        Replace the values in square brackets with the appropriate values. This will take a few hours to run.
        The last <code>--vcfs</code> option is optional and only required if you want to add additional variants to the
        simluation.
        It is recommended to run GATK LeftAlignAndTrimVariants on any VCF file that is used as input.
    </p>

<pre>
./varsim.py --vc_in_vcf All.vcf.gz --sv_insert_seq insert_seq.txt \
--sv_dgv GRCh37_hg19_supportingvariants_2013-07-23.txt \
--reference hs37d5.fa --id simu --read_length 100 --vc_num_snp 3000000 --vc_num_ins 100000 \
--vc_num_del 100000 --vc_num_mnp 50000 --vc_num_complex 50000 --sv_num_ins 2000 \
--sv_num_del 2000 --sv_num_dup 200 --sv_num_inv 1000 --sv_percent_novel 0.01 \
--vc_percent_novel 0.01 --mean_fragment_size 350 --sd_fragment_size 50 \
--vc_min_length_lim 0 --vc_max_length_lim 49 --sv_min_length_lim 50 \
--sv_max_length_lim 1000000 --nlanes 5 --total_coverage 1 \
--simulator_executable ART/art_bin_VanillaIceCream/art_illumina --out_dir out --log_dir log --work_dir work \
--simulator art  \
--vcfs [Optional VCF file to include, variants.vcf]
</pre>

    <p>
        The reads will be generated in the <code>out</code> directory. A script <code>quickstart.sh</code> is also
        provided which
        runs the above steps and generates reads for 1X coverage. Run the secondary analysis tools (alignment and
        variant calling) on those.
        The ground truth VCF file is also in the <code>out</code> directory called <code>simu.truth.vcf</code>
    </p>

    <p>
        <b>Step 6:</b> After running the alignment and variant calling we can evaluate the results. In order to validate
        the variants run the following command:
    </p>

<pre>
java -jar VarSim.jar vcfcompare -true_vcf simu.truth.vcf -prefix simu [VCF from result of secondary analysis] 
</pre>

    <p>
        This will output a JSON file that can be used as input to the VCF Compare webapp [<a
            href="http://bioinform.github.io/varsim/webapp/variant_compare.html"
            onclick="trackOutboundLink('webapp/variant_compare.html'); return false;">http://bioinform.github.io/varsim/webapp/variant_compare.html</a>]
    </p>
    <p>NOTE: for better results, please use compare_vcf.py which leverages both VarSim and RTG vcfeval.
    </p>

    <p>
        <b>Step 7:</b> In order to validate the alignments run the following command:
    </p>

<pre>
java -jar VarSim.jar samcompare -prefix simu [BAM files from result of secondary analysis]
</pre>

    <p>
        This will output a JSON file that can be used as input to the Alignment Compare webapp [<a
            href="http://bioinform.github.io/varsim/webapp/alignment_compare.html"
            onclick="trackOutboundLink('webapp/alignment_compare.html'); return false;">http://bioinform.github.io/varsim/webapp/alignment_compare.html</a>]
    </p>


    <h2>Quick Start Guide for Somatic Simulation [in writing]</h2>
    <p>
        <b>Step 1:</b> Make sure you have already gone through the Quick Start Guide for germline simulation
    </p>

    <p>
        <b>Step 2:</b> Currently VarSim can only simulate random sets of variants from the COSMIC database.
        This limitation will be addressed in a future version. Hence, this step will be to acquire the COSMIC
        database VCF.
    </p>

    <p>
        Download two two cosmic VCF files from http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/download .
        <code>CosmicCodingMuts.vcf.gz</code> and <code>CosmicNonCodingVariants.vcf.gz</code>
    </p>

    <p>
        After downloading the 2 cosmic VCFs to the directory <code>varsim_run</code> run the following command:
    </p>

    <pre>
        cat <(gzip -dc CosmicCodingMuts.vcf.gz) <(gzip -dc CosmicNonCodingVariants.vcf.gz) | gzip -c > cosmic.vcf.gz
    </pre>

    <p>
        This will make a concatentated VCF. Although this is not a valid VCF file, VarSim is ok with it.
        You can also provide your own COSMIC VCF file too. Just make sure it is a single file. VarSim uses the "COS"
        in the cosmic SNP annotations to separate them out. This will change in future versions.
    </p>

    <p>
        <b>Step 3:</b> Simulate tumor reads and variants by running the following command:
    </p>

    <pre>
        varsim_somatic.py --reference hs37d5.fa --id cosmic --som_num_snp 10000 \
        --som_num_ins 2000 --som_num_del 2000 \
        --som_num_mnp 200 \
        --som_num_complex 200 \
        --cosmic_vcf cosmic.vcf.gz \
        --normal_vcf out/simu.truth.vcf \
        --nlanes 5 --total_coverage 1 \
        --simulator art \
        --simulator_executable ART/art_bin_VanillaIceCream/art_illumina \
        --out_dir som_out --log_dir som_log --work_dir som_work &> somatic.log
    </pre>

    <p>
        <b>Step 4 [optional]:</b>  Mix in some normal reads into the tumor. In order to roughly simulate normal
        contamination you can take advantage of the multiple lanes easily mix the normal and tumor reads.
    </p>

    <p>
        For example, in order to simulate reads with 50x coverage of a normal sample and 50x coverage of
        a tumor sample with tumor allele frequency of 0.3. One possible way to achieve this is to
        simulate 7 lanes of normal sample at 70x coverage and 3 lanes of tumor sample at 30x coverage.
        Take the first 5 lanes of the simulated normal sample as the pure normal. Then take the remaining
        two lanes of normal with all the simulated tumor samples lanes as the contaminated tumor.
    </p>

    <p>
        <b>Step 5:</b> Run somatic analysis pipeline with the simulated reads
    </p>

    <p>
        <b>Step 6:</b> Validate alignments and variants called in the same way as before for the germline case.
        The truth VCF is in the file <code>som_out/cosmic_somatic.vcf</code>.
    </p>

    <h2>Command Line Options</h2>
<pre>
	VarSim: A high-fidelity simulation validation framework

optional arguments:
  -h, --help            show this help message and exit
  --out_dir DIR         Output directory for the simulated genome, reads and
                        variants (default: out)
  --work_dir DIR        Work directory, currently not used (default: work)
  --log_dir DIR         Log files of all steps are kept here (default: log)
  --reference FASTA     Reference genome that variants will be inserted into
                        (default: None)
  --seed seed           Random number seed for reproducibility (default: 0)
  --sex Sex             Sex of the person (MALE/FEMALE) (default: MALE)
  --id ID               Sample ID to be put in output VCF file (default: None)
  --simulator SIMULATOR
                        Read simulator to use (default: art)
  --varsim_jar PATH     Path to VarSim.jar (default: /net/kodiak/volumes/river
                        /shared/users/johnmu/github/varsim/VarSim.jar)
  --read_length LENGTH  Length of read to simulate (default: 100)
  --nlanes INTEGER      Number of lanes to generate, coverage will be divided
                        evenly over the lanes. Simulation is parallized over
                        lanes. Each lane will have its own pair of files
                        (default: 1)
  --total_coverage FLOAT
                        Total coverage to simulate (default: 1.0)
  --mean_fragment_size FLOAT
                        Mean fragment size to simulate (default: 350)
  --sd_fragment_size FLOAT
                        Standard deviation of fragment size to simulate
                        (default: 50)
  --vcfs VCF [VCF ...]  Addtional list of VCFs to insert into genome, priority
                        is lowest ... highest (default: [])
  --force_five_base_encoding
                        Force output bases to be only ACTGN (default: False)
  --filter              Only use PASS variants for simulation (default: False)
  --keep_temp           Keep temporary files after simulation (default: False)

Pipeline control options. Disable parts of the pipeline.:
  --disable_rand_vcf    Disable sampling from the provided small variant VCF
                        (default: False)
  --disable_rand_dgv    Disable sampline from the provided DGV file (default:
                        False)
  --disable_vcf2diploid
                        Disable diploid genome simulation (default: False)
  --disable_sim         Disable read simulation (default: False)

Small variant simulation options:
  --vc_num_snp INTEGER  Number of SNPs to sample from small variant VCF
                        (default: 0)
  --vc_num_ins INTEGER  Number of insertions to sample from small variant VCF
                        (default: 0)
  --vc_num_del INTEGER  Number of deletions to sample from small variant VCF
                        (default: 0)
  --vc_num_mnp INTEGER  Number of MNPs to sample from small variant VCF
                        (default: 0)
  --vc_num_complex INTEGER
                        Number of complex variants to sample from small
                        variant VCF (default: 0)
  --vc_percent_novel FLOAT
                        Percent variants sampled from small variant VCF that
                        will be moved to novel positions (default: 0)
  --vc_min_length_lim INTEGER
                        Min length of small variant to accept [inclusive]
                        (default: 0)
  --vc_max_length_lim INTEGER
                        Max length of small variant to accept [inclusive]
                        (default: 99)
  --vc_in_vcf VCF       Input small variant VCF, usually dbSNP (default: None)
  --vc_prop_het FLOAT   Proportion of heterozygous small variants (default:
                        0.6)

Structural variant simulation options:
  --sv_num_ins INTEGER  Number of insertions to sample from DGV (default: 20)
  --sv_num_del INTEGER  Number of deletions to sample from DGV (default: 20)
  --sv_num_dup INTEGER  Number of duplications to sample from DGV (default:
                        20)
  --sv_num_inv INTEGER  Number of inversions to sample from DGV (default: 20)
  --sv_percent_novel FLOAT
                        Percent variants sampled from DGV that will be moved
                        to novel positions (default: 0)
  --sv_min_length_lim min_length_lim
                        Min length of structural variant to accept [inclusive]
                        (default: 100)
  --sv_max_length_lim max_length_lim
                        Max length of structural variant to accept [inclusive]
                        (default: 1000000)
  --sv_insert_seq FILE  Path to file containing concatenation of real
                        insertion sequences (default: None)
  --sv_dgv DGV_FILE     DGV file containing structural variants (default:
                        None)

DWGSIM options:
  --dwgsim_start_e first_base_error_rate
                        Error rate on the first base (default: 0.0001)
  --dwgsim_end_e last_base_error_rate
                        Error rate on the last base (default: 0.0015)
  --dwgsim_options DWGSIM_OPTIONS
                        DWGSIM command-line options (default: )

ART options:
  --profile_1 profile_file1
                        ART error profile for first end (default: None)
  --profile_2 profile_file2
                        ART error profile for second end (default: None)
  --art_options ART_OPTIONS
                        ART command-line options (default: )
</pre>


    <h2>References/Tools</h2>

    <ul style="font-family:monospace;">
        <li>vcf2diploid: Rozowsky J, Abyzov A, Wang J, Alves P, Raha D, Harmanci A, Leng J, Bjornson R, Kong Y,
            Kitabayashi N, Bhardwaj N, Rubin M, Snyder M, Gerstein M. <a
                    href="http://msb.embopress.org/content/7/1/522">AlleleSeq: analysis of allele-specific expression
                and binding in a network framework</a>. Mol Syst Biol. 2011 Aug 2;7:522. doi: 10.1038/msb.2011.54. <a
                    href="http://alleleseq.gersteinlab.org/tools.html">Download link</a></li>
        <li>DWGSIM: Nils Homer. <a href="https://github.com/nh13/DWGSIM">Github repository</a></li>
        <li>ART: Huang W1, Li L, Myers JR, Marth GT. <a href="http://www.ncbi.nlm.nih.gov/pubmed/22199392">ART: a
            next-generation sequencing read simulator.</a> Bioinformatics. 2012 Feb 15;28(4):593-4. doi:
            10.1093/bioinformatics/btr708. Epub 2011 Dec 23. <a
                    href="http://www.niehs.nih.gov/research/resources/software/biostatistics/art/">Download link</a>
        </li>
    </ul>


</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
</body>
